File: Python/Binary Tree Level Order Traversal II.py
# Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).

# For example:
# Given binary tree {3,9,20,#,#,15,7},
    # 3
   # / \
  # 9  20
    # /  \
   # 15   7
# return its bottom-up level order traversal as:
# [
  # [15,7]
  # [9,20],
  # [3],
# ]


# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def levelOrderBottom(self, root):
		if root == None:
			return []
		queue = []
		result = []
		result.append([root.val])
		queue.append(root.left)
		queue.append(root.right)
		while len(queue) != 0:
			currentList = []
			currentlen = len(queue)
			for i in range(0,currentlen):
				temp = queue.pop(0)
				if temp != None:
					queue.append(temp.left)
					queue.append(temp.right)
					currentList.append(temp.val)
			if len(currentList) != 0 :
			    result.insert(0,currentList)
		return result

File: Java/Binary Tree Level Order Traversal II.java
class Binary Tree Level Order Traversal II{
	public ArrayList<ArrayList<Integer>> levelOrderBottom(TreeNode root) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if(root == null)
        	return res;
        LinkedList<TreeNode> queue = new LinkedList<TreeNode>();
        queue.offer(root);
        while(!queue.isEmpty()){
        	int size = queue.size();
        	ArrayList<TreeNode> temp = new ArrayList<TreeNode>();
        	for(int i=0;i<size;i++){
        		temp.add(queue.poll());
        	}
        	ArrayList<Integer> sol = new ArrayList<Integer>();
        	for(TreeNode t:temp){
        		sol.add(t.val);
        		if(t.left != null)
        			queue.offer(t.left);
        		if(t.right != null)
        			queue.offer(t.right);
        	}
        	res.add(sol);
        }
        Collections.reverse(res);
        return res;
    }
}